### PR TITLE
add biome change per player

### LINKED ISSUE #9889: Change biomes per-player in chunk/biome packets
### Is your feature request related to a problem?

There isn't a way to set biomes to be per-player

### Describe the solution you'd like.

In the `PlayerChunkLoadEvent` or a new superevent of that, there should be a way to change the biome data sent to the client in order for each player to be able to get a different set of biomes.

I think this has to be a new superevent of PlayerChunkLoadEvent because biomes can change independent of chunks being sent to the client with the new fillbiome command. There is also a specific packet for changing a chunk biomes.

### Describe alternatives you've considered.

N/A

### Other

I'm not sure of the correct API structure for representing chunk biomes. Generally, I think it would be beneficial if plugins could create chunk biome data arbitrarily to they aren't creating new PalettedContainers for every player. Having some way plugins could create biome data, cache it, and then use it for specific players would be ideal.

ChunkSnapshot is a good place to start. It wraps around 2 arrays of PalettedContainers for blockstates and biomes. Having a supertype of that which just wraps biomes seems like a good idea.

Something to note is that the array of PalettedContainer for biomes has a different length depending on the height of the world. So care would have to be taken to ensure that the sizes matched.

### PR DESCRIPTION
**Pull request by yannicklamprecht**
resolves #9889 

pre hard fork: https://github.com/PaperMC/Paper/pull/10284

# TODO:

- [ ] Implement and call event
- [ ] finish implementing BiomesSnapshot
- [ ] add a handle to Chunk
- [ ] test api with plugin
- [ ] test compatibility with plugins using PlayerChunkLoadEvent
- [ ] reduce diff
- [ ] add paper comments

### CODE REVIEW

File: paper-api/src/main/java/io/papermc/paper/event/packet/PlayerBiomesLoadEvent.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.bukkit.Chunk;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.world.ChunkEvent;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * Is called when a {@link Player} receives {@link org.bukkit.block.Biome}s
```
Comment (ShaneBeee) [inline paper-api/src/main/java/io/papermc/paper/event/packet/PlayerBiomesLoadEvent.java]:
```suggestion
 * Is called when a {@link Player} receives {@link org.bukkit.block.Biome Biomes}
```
Looks nicer in the javadocs

Exhibit A:
Screenshot 2024-12-24 at 7 04 27 PM

Comment (yannicklamprecht) [reply paper-api/src/main/java/io/papermc/paper/event/packet/PlayerBiomesLoadEvent.java]: The plural doesn't seem to work then. At least what I can see in your screen shoot.

Comment (ShaneBeee) [reply paper-api/src/main/java/io/papermc/paper/event/packet/PlayerBiomesLoadEvent.java]: Screenshot 2024-12-25 at 11 25 40 AM
it puts the "s" as part of the word/link

Comment (yannicklamprecht) [reply paper-api/src/main/java/io/papermc/paper/event/packet/PlayerBiomesLoadEvent.java]: Thank you, changed it.

**Comment by yannicklamprecht** [comment]
Nothing that I'll finish in near future.

**Pull request closed without merging**